Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp

```diff
@@ -531,22 +531,26 @@ void Recorder::drawControlPart(ControlPart& part, const FloatRoundedRect& border
 
 void Recorder::clip(const FloatRect& rect)
 {
+    appendStateChangeItemIfNecessary(); // Conservative: we do not know if the clip application might use state such as antialiasing.
```

Comment:
New calls have been added recently. I think we have 63 display list items. Currently, there are 36 calls to this function. By adding 7 calls by this PR, we will have 43 calls.

I think we should be conservative in calling this function. How about adding a new function like this:

```
void Recorder::appendStateChangeItemIfNecessary(GraphicsContextState::Change change)
```
or
```
void Recorder::appendStateChangeItemIfNecessary(GraphicsContextState::ChangeFlags changes)
```
This function will call the existing one if the current state changes the context with `change` or any of `changes`. The above function can call it like this
```
appendStateChangeItemIfNecessary(GraphicsContextState::Change::ShouldAntialias);
```

Reply:
So `appendStateChangeItemIfNecessary()` of course already consults the change flags and only sends state changes that have actually occurred.

> I think we have 63 display list items. Currently, there are 36 display list items call to this function. New calls to this function have been added recently. By adding 7 calls by this PR, we will have 43 calls.

In general, almost all calls should sync the state. I don't think there's many calls that draw that do not need the exactly correct draw state. To me it just sounds natural and I'm not terribly worried the increase of calls. I'm more worried about the calls that don't have the state sync -- why are they correct?? 

If I understand correctly, you're concerned that if there are pending change tags, say, `GraphicsContextState::Change::ShouldAntialias` and `GraphicsContextState::Change::ShouldSmoothFonts`, and `clip()` only needs `ShouldAntialias`, then we shouldn't send `ShouldSmoothFonts `?

I'm not sure that is an optimisation at all. Generally the `appendStateChangeItemIfNecessary` is an optimisation for batching consecutive changes to the state. In other words: the painting, typically, does not set state that is not needed. Thus for CSS painting, vast majority of the state that is set is also used. 

Sending all the set state in itself is really lightweight operation.

Canvas could change state that is not used, but that, too, never have shown up in our benchmarks -- this function is does not seem to do redundant work. We also don't have that optimisation for non-gpup: current in-process context does set all requested state eagerly, even  though the benefit should be the same for gpu and non-gpu.

If the state is set, it is likely to be used. There's no evidence that we set the state redundantly. This would mean that more fine-grained state setting would be just splitting the state setting into two state setting functions, something that probably in itself is a regression (two IPC commands vs one). The only gain would be if some state is set redundantly, and we'd actually skip using it altogether.

It is also very easy to get wrong -- very much attention would need to be put to re-learn all commands and which state they depend on.

Maybe making that change would need some data to back it as an optimisation?

Reply:
> If I understand correctly, you're concerned that if there are pending change tags, say, GraphicsContextState::Change::ShouldAntialias and GraphicsContextState::Change::ShouldSmoothFonts, and clip() only needs ShouldAntialias, then we shouldn't send ShouldSmoothFonts ?

No I did not mean that. Consider these two cases:

// Case 1: context state does not have ShouldAntialias change
```
context.setFillColor(...)
context.clip(...)
```

// Case 2: context state does have ShouldAntialias change
```
context.setShouldAntialias(...)
context.clip(...)
```

For case 1, we may not need to `appendStateChangeItemIfNecessary()` because the pending context state does not have anything that will affect clipping. However for case 2, we need to commit the pending context state before applying the clipping.

So if we make `Recorder::clip()` calls 

```
appendStateChangeItemIfNecessary(GraphicsContextState::Change::ShouldAntialias);
```
It will not call `appendStateChangeItemIfNecessary()` for case 1 but will call it for case 2.

Reply:
Pleas note that my suggestion is to make DisplayList::Recorder have two versions of `appendStateChangeItemIfNecessary()`; something like that:
```
void appendStateChangeItemIfNecessary();
void appendStateChangeItemIfNecessary(GraphicsContextState::ChangeFlags changes);
```